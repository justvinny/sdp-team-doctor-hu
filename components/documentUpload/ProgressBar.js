import React, { useEffect, useState, useLayoutEffect, useContext, Component} from "react";
import { StyleSheet, TouchableOpacity, View,} from 'react-native';
import Image from 'react-native-image-progress';
import * as Progress from 'react-native-progress';

const ProgressBar = () => {
  
    return (

        <Progress.Bar
           style={styles.progress}
            indeterminate= {true}
            progress = {0}
            width={250}
            color={'#8870e6'}
          
        />
    );
  
};


export default  ProgressBar ;

const styles = StyleSheet.create({
    container: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: '#fff',
      paddingVertical: 20,
    },
    progress: {
      margin: 10,
      
    },
  });