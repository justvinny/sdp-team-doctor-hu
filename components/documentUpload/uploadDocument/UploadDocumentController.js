// import React in our code
import React, { useEffect, useState } from "react";
// import all the components we are going to use
import { Platform, Alert} from "react-native";
import * as ImagePicker from "expo-image-picker";
import { storage } from "../../../firebase/firebaseConfig";
import firestoreService from "../../../firebase/firestoreService";
import * as DocumentPicker from "expo-document-picker";
import Staff from "../../../models/Staff";
import UploadDocumentView from "./UploadDocumentView";

// Upload documents/images to patient profile
function UploadDocumentController({
  toggleDocumentOverlay,
  patient,
  staff,
  patientName,
}) {
  const [file, setFile] = useState("");
  const [download, showDownload] = useState(false);
  const [title, setTitle] = useState("");
  const [progress, showProgress] = useState(false);

  if (Platform.OS === "android" || Platform.OS === "ios") {
    useEffect(() => {
      const LogBox = require("react-native").LogBox;
      LogBox.ignoreLogs(["Animated: `useNativeDriver`"]);
    }, []);
  }

  // can upload either a camera image or doucment
  //image picker
  const imagePicker = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: false,
    });

    if (!result.cancelled) {
      setFile(result.uri);
      showDownload(true);
    } else {
      showDownload(false);
    }
  };

  //document picker
  const pickDocument = async () => {
    let result = await DocumentPicker.getDocumentAsync({
      type: "*/*",
      copyToCacheDirectory: false,
    });

    if (result.type !== "cancel") {
      setFile(result.uri);
      showDownload(true);

      if (result?.name !== undefined) {
        Alert.alert(
          "Use the existing document title or make your own?",
          "Exisiting title: " + result.name,
          [
            {
              text: "No",
              style: "cancel",
            },
            {
              text: "Yes",
              onPress: () => {
                setTitle(result.name);
              },
              style: "destructive",
            },
          ]
        );
      }
    } else {
      showDownload(false);
    }
  };

  // uploads document to firestorage and firebase
  const upload = async () => {
    if (file) {
      try {
        showProgress(true);
        const childPath = `document/${patient}/${Math.random().toString(36)}`;
        const response = await fetch(file);
        const blob = await response.blob();
        const task = await storage.ref().child(childPath).put(blob);

        task.ref.getDownloadURL().then((url) => {
          const newDocument = {
            staffId: staff,
            patientId: patient,
            url: url,
            timestamp: Date.now(),
            title: title,
          };

          //adds upload doucment to patient and staff profiles
          firestoreService.addMedicalResult(staff, newDocument);
          firestoreService.addMedicalResult(patient, newDocument);

          showProgress(false);
          showDownload(false);

          // Send notification to user
          firestoreService.getUserById(staff).then((user) => {
            const _staff = Staff.staffFirestoreFactory(user);
            const newNotification = {
              type: "result",
              content: newDocument.title,
              isRead: false,
              timestamp: newDocument.timestamp,
              from: _staff.getFullName(),
            };
            firestoreService.addNotification(
              newDocument.patientId,
              newNotification
            );
          });

          Alert.alert(
            "Document '" + title + "' for " + patientName,
            "Document updated successfully!",
            [
              {
                text: "Thanks!",
              },
            ]
          );
          setTitle("");
        });
      } catch (error) {
          showProgress(false);
          showDownload(false);
        alert(error.message);
      }
    }
  };

  useEffect(() => {
    (async () => {
      if (Platform.OS !== "web") {
        const { status } =
          await ImagePicker.requestMediaLibraryPermissionsAsync();
        if (status !== "granted") {
          alert("Sorry we need camera roll permissions to make this work!");
        }
      }
    })();
  }, []);

  // checks title is present into order to progresses to upload
  const checkTitleInput = () => {
    //Check for the title TextInput
    if (!title.trim()) {
      alert("Please enter a Document Title before uploading.");
      return;
    } else {
      upload();
    }
  };

  return <UploadDocumentView 
    toggleDocumentOverlay={toggleDocumentOverlay}
    patientName={patientName}
    download={download}
    title={title}
    setTitle={setTitle}
    progress={progress}
    pickDocument={pickDocument}
    imagePicker={imagePicker}
    checkTitleInput={checkTitleInput}
  />
}

export default UploadDocumentController;

