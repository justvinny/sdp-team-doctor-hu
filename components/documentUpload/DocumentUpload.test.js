import React from "react";
import { render  } from "@testing-library/react-native";
import UploadDocumentController from "./uploadDocument/UploadDocumentController";
import ProgressBar from "./uploadDocument/ProgressBar";


describe('<UploadDocumentController />', () => {
    const tree = render(<UploadDocumentController />).toJSON();
    it('UploadDocumentController Component renders correctly', () => {
        expect(tree).toMatchSnapshot();
    });
});
describe('<ProgressBar />', () => {
    const tree = render(<ProgressBar />).toJSON();
    it('ProgressBar Component renders correctly', () => {
        expect(tree).toMatchSnapshot();
    });
});