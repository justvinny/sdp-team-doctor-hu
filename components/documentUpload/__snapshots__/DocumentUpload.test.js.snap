// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> ProgressBar Component renders correctly 1`] = `
<View
  animated={true}
  animationConfig={
    Object {
      "bounciness": 0,
    }
  }
  animationType="spring"
  indeterminate={true}
  indeterminateAnimationDuration={1000}
  onLayout={[Function]}
  progress={0}
  style={
    Array [
      Object {
        "backgroundColor": undefined,
        "borderColor": "#8870e6",
        "borderRadius": 4,
        "borderWidth": 1,
        "overflow": "hidden",
        "width": 250,
      },
      Object {
        "margin": 10,
      },
    ]
  }
  useNativeDriver={false}
>
  <View
    style={
      Object {
        "backgroundColor": "#8870e6",
        "height": 6,
        "transform": Array [
          Object {
            "translateX": -74.39999999999999,
          },
          Object {
            "translateX": -86.80000000000001,
          },
          Object {
            "scaleX": 0.30007,
          },
        ],
      }
    }
  />
</View>
`;

exports[`<UploadDocumentController /> UploadDocumentController Component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "padding": 15,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Upload Document for:
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
       
    </Text>
  </Text>
  <View
    style={
      Object {
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChangeText={[Function]}
        placeholder="Document Title"
        placeholderTextColor="#86939e"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#242424",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "margin": 5,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 10,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 30,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Choose Image
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 10,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 30,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Choose Document
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "red",
            "borderColor": "#2089dc",
            "borderRadius": 10,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 30,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Cancel
        </Text>
      </View>
    </View>
  </View>
</View>
`;
